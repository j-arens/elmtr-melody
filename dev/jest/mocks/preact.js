module.exports = {
    unmountComponent: jest.fn(),
};
