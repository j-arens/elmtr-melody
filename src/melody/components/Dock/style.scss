.dock {
    position: relative;

    &__toggle {
        display: flex;
        justify-content: center;
        align-items: center;
    }

    &__toggleIcon {
        width: 20px;
        height: 20px;
        fill: currentColor;
    }

    &__arrow {
        position: absolute;
        top: 100%;
        right: 4px;
        width: 12px;
        height: 12px;
        background-color: $color_mine_shaft;
        border-top: 1px solid #000;
        border-left: 1px solid #000;
        pointer-events: none;
        transform: rotate(45deg);
        z-index: 2;
        box-shadow: 1px 2px 3px 1px rgba(30, 30, 30, 0.5);
    }

    &__controlsGroup {
        position: absolute;
        top: calc(100% + 6px);
        right: -11px;
        background-color: $color_mine_shaft;
        border: 1px solid #000;
        border-radius: 2px;
        box-shadow: 1px 2px 2px 1px rgba(30, 30, 30, 0.2);
        z-index: 1;
    }

    &__control {
        display: flex;
        align-items: center;
        font-family: sans-serif;
        color: $color_white !important;
        font-weight: 400;
        font-size: 12px;
        width: 100%;
        padding: 8px 6px 8px 4px;
        border-radius: 0;
        position: relative;
        cursor: pointer;

        &::before {
            content: '';
            width: calc(1.25em + 8px);
            height: 100%;
            position: absolute;
            top: 0;
            left: 0;
            background-color: $color_tundora;
        }

        &:active {
            transform: none;
            opacity: 0.9;
        }

        &--disabled {
            cursor: not-allowed;
            color: #ccc !important;
            opacity: 0.5;

            &:hover {
                color: currentColor;
                opacity: 0.5;
            }

            &:active {
                transform: none;
                opacity: 0.5;
                color: currentColor;
            }
        }
    }

    &__control + &__control {
        border-top: 1px solid #000;
    }

    &__controlIcon {
        fill: $color_white;
        width: 1.25em;
        height: 1.25em;
        margin-right: 11px;
        position: relative;
        z-index: 1;
    }

    &__controlName {
        display: block;
        white-space: nowrap;
    }
}