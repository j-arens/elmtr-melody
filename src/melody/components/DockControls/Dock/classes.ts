import { prefixClasses } from '@utils/index';

export const baseBtns = (prefix: string) => prefixClasses(
    `elementor-element-${prefix}-`,
    'dock-controls-secondary-color',
    'dock-controls-text-color',
    'dock-controls-font',
);

export const dockControls = (prefix: string) => prefixClasses(
    `elementor-element-${prefix}-`,
    'dock-controls-primary-color',
);

export const dockArrow = (prefix: string) => prefixClasses(
    `elementor-element-${prefix}-`,
    'dock-controls-arrow-color',
    'dock-controls-border-color',
);

export const controlsGroup = (prefix: string) => prefixClasses(
    `elementor-element-${prefix}-`,
    'dock-controls-border-color',
);

export const controlIcon = (prefix: string) => prefixClasses(
    `elementor-element-${prefix}-`,
    'dock-controls-icon-color',
);
