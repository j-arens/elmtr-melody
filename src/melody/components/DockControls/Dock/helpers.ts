import { DockToggleDims } from '@melody/redux/modules/ui/type';

export interface DockCoordinates {
    top: number;
    right: number;
}

const X_PADDING = 12;
const Y_PADDING = 12;

export function dockPosition(dims: DockToggleDims): DockCoordinates {
    const top = dims.y + Y_PADDING;
    let right = (window.innerWidth - dims.x) - dims.width - X_PADDING;

    if (right > window.innerWidth) {
        right = window.innerWidth - X_PADDING;
    }

    return {
        top,
        right,
    };
}
