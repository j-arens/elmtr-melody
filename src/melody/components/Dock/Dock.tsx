import BaseButton from '@components/BaseButton/';
import Icon from '@components/Icon/';
import { WithOptionalClassName } from '@melody/components/type';
import { Action, Track } from '@redux/type';
import { NO_OP } from '@utils/index';
import { __ } from '@wordpress/i18n';
import * as classnames from 'classnames';
import { h } from 'preact';
const s = require('./style.scss');

export interface StateProps {
    showDock: boolean;
    track: Track;
    playbackRate: number;
}

export interface DispatchProps {
    toggleDock: () => Action;
    speedUp: () => Action;
    slowDown: () => Action;
}

export interface OwnProps extends WithOptionalClassName {
    triggerClassName?: string;
}

type Props = StateProps & DispatchProps & OwnProps;

export default ({
    showDock,
    toggleDock,
    track,
    speedUp,
    slowDown,
    playbackRate,
    className = '',
    triggerClassName = '',
}: Props) => {
    const handleDownload = () => {
        const { download_url, attributes: { origin } } = track;
        if (origin === 'internal') {
            window.location.href = download_url;
        } else {
            window.open(download_url);
        }
    };
    const speedUpClasses = classnames(s.dock__control, {
        [s['dock__control--disabled']]: playbackRate === 2,
    });
    const slowDownClasses = classnames(s.dock__control, {
        [s['dock__control--disabled']]: playbackRate === 0.5,
    });
    return (
        <div class={`${s.dock} ${className}`}>
            <BaseButton
                onClick={toggleDock}
                className={`${s.dock__toggle} ${triggerClassName}`}
            >
                <Icon className={s.dock__toggleIcon} name="settings" />
            </BaseButton>
            {showDock && [
                <span key="arrow" class={s.dock__arrow} />,
                <div key="group" class={s.dock__controlsGroup}>
                    {track.download_url &&
                        <BaseButton
                            onClick={handleDownload}
                            className={s.dock__control}
                        >
                            <Icon className={s.dock__controlIcon} name="download" />
                            <span class={s.dock__controlName}>{__('Download')}</span>
                        </BaseButton>
                    }
                    <BaseButton
                        onClick={playbackRate === 2 ? NO_OP : speedUp}
                        className={speedUpClasses}
                    >
                        <Icon className={s.dock__controlIcon} name="plus" />
                        <span class={s.dock__controlName}>{__('Speed up')}</span>
                    </BaseButton>
                    <BaseButton
                        onClick={playbackRate === 0.5 ? NO_OP : slowDown}
                        className={slowDownClasses}
                    >
                        <Icon className={s.dock__controlIcon} name="minus" />
                        <span class={s.dock__controlName}>{__('Slow down')}</span>
                    </BaseButton>
                </div>]
            }
        </div>
    );
};
