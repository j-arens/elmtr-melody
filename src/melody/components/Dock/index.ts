import {
    slowDown,
    speedUp,
    toggleDock,
} from '@redux/actions';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import Dock, { DispatchProps, OwnProps, StateProps } from './Dock';

const mapStateToProps = (state: State) => ({
    showDock: state.ui.showDock,
    track: state.tracks[state.currentTrack],
    playbackRate: state.playbackRate,
});

const mapDispatchToProps = dispatch => ({
    toggleDock: () => dispatch(toggleDock()),
    speedUp: () => dispatch(speedUp()),
    slowDown: () => dispatch(slowDown()),
});

export default connect<
    StateProps,
    DispatchProps,
    OwnProps
>(mapStateToProps, mapDispatchToProps)(Dock);
