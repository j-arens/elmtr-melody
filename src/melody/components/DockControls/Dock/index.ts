import { slowDown, speedUp } from '@redux/modules/audio/actions';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import Dock, { DispatchProps, StateProps } from './Dock';

const mapStateToProps = (state: State) => ({
    showDock: state.ui.showDock,
    track: state.audio.tracks[state.audio.currentTrack],
    playbackRate: state.audio.playbackRate,
    coordinates: state.ui.dockCoordinates,
    wrapperId: state.ui.wrapperId,
});

const mapDispatchToProps = dispatch => ({
    speedUp: () => dispatch(speedUp()),
    slowDown: () => dispatch(slowDown()),
});

export default connect<
    StateProps,
    DispatchProps,
    {}
>(mapStateToProps, mapDispatchToProps)(Dock);
