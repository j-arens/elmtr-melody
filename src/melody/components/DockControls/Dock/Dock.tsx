import BaseButton from '@components/BaseButton/';
import Icon from '@components/Icon/';
import Portal from '@components/Portal/Portal';
import { WithOptionalClassName } from '@melody/components/type';
import { GLOBAL } from '@melody/constants';
import { DockToggleDims } from '@melody/redux/modules/ui/type';
import { Action, Track } from '@redux/type';
import { cySelector, NO_OP } from '@utils/index';
import * as classnames from 'classnames';
import { Component, h } from 'preact';
import * as classes from './classes';
import { dockPosition } from './helpers';
const s = require('../style.scss');

const { jQuery: $ } = GLOBAL;

export interface StateProps {
    showDock: boolean;
    track: Track;
    playbackRate: number;
    coordinates: DockToggleDims;
    wrapperId: string;
}

export interface DispatchProps {
    speedUp: () => Action;
    slowDown: () => Action;
}

interface State {
    width: number;
}

type Props = StateProps & DispatchProps & WithOptionalClassName;

export default class extends Component<Props, State> {
    static defaultProps = {
        className: '',
        customizationClasses: '',
    };

    state = {
        width: null,
    };

    el: HTMLElement;

    componentDidMount() {
        if (this.el) {
            this.setState({
                width: this.el.offsetWidth,
            });
        }
    }

    handleDownload = () => {
        const { track: { download_url } } = this.props;
        window.open(download_url, '_blank');
    }

    portalDock(inner) {
        const { wrapperId } = this.props;
        const el = document.getElementById(`melody-widgetRoot:${wrapperId}`);
        const node = $(el).closest('.elementor-section').get(0);
        return (
            <Portal into={node}>
                {inner}
            </Portal>
        );
    }

    render(props: Props, { width }: State) {
        const {
            showDock,
            track,
            speedUp,
            slowDown,
            playbackRate,
            coordinates,
            className,
            wrapperId,
        } = props;

        const baseBtnClasses = classnames(
            s.dock__control,
            'elementor-element',
            classes.baseBtns(wrapperId),
        );

        const speedUpClasses = classnames(baseBtnClasses, {
            [s['dock__control--disabled']]: playbackRate === 2,
        });

        const slowDownClasses = classnames(baseBtnClasses, {
            [s['dock__control--disabled']]: playbackRate === 0.5,
        });

        const { controls, arrow } = dockPosition(coordinates, width);

        const dockControls = (
            <div
                ref={el => this.el = el as HTMLElement}
                class={classnames(
                    s.dock,
                    'elementor-element',
                    className,
                    classes.dockControls(wrapperId),
                )}
                style={{
                    pointerEvents: showDock ? 'all' : 'none',
                    visibility: showDock ? 'visible' : 'hidden',
                    ...controls,
                }}
                {...cySelector('dock-controls')}
            >
                <span
                    class={classnames(
                        s.dock__arrow,
                        'elementor-element',
                        classes.dockArrow(wrapperId),
                    )}
                    style={arrow}
                />
                <div
                    class={classnames(
                        s.dock__controlsGroup,
                        'elementor-element',
                        classes.controlsGroup(wrapperId),
                    )}
                    {...cySelector('dock-controls-wrap')}
                >
                    {track.download_url &&
                        <BaseButton
                            onClick={this.handleDownload}
                            className={baseBtnClasses}
                            {...cySelector('download')}
                        >
                            <Icon
                                className={classnames(
                                    s.dock__controlIcon,
                                    'elementor-element',
                                    classes.controlIcon(wrapperId),
                                )}
                                name="download"
                            />
                            <span class={s.dock__controlName}>Download</span>
                        </BaseButton>
                    }
                    <BaseButton
                        onClick={playbackRate === 2 ? NO_OP : speedUp}
                        className={speedUpClasses}
                        {...cySelector('speed-up')}
                    >
                        <Icon
                            className={classnames(
                                s.dock__controlIcon,
                                'elementor-element',
                                classes.controlIcon(wrapperId),
                            )}
                            name="plus"
                        />
                        <span class={s.dock__controlName}>Speed up</span>
                    </BaseButton>
                    <BaseButton
                        onClick={playbackRate === 0.5 ? NO_OP : slowDown}
                        className={slowDownClasses}
                        {...cySelector('slow-down')}
                    >
                        <Icon
                            className={classnames(
                                s.dock__controlIcon,
                                'elementor-element',
                                classes.controlIcon(wrapperId),
                            )}
                            name="minus"
                        />
                        <span class={s.dock__controlName}>Slow down</span>
                    </BaseButton>
                </div>
            </div>
        );

        if (process.env.NODE_ENV !== 'test') {
            return this.portalDock(dockControls);
        }

        return dockControls;
    }
}
