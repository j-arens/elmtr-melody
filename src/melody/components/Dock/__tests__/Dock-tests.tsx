import { h } from 'preact';
import { render } from 'preact-render-to-string';
import Dock from '../Dock';

describe('Dock', () => {
    let track;

    beforeEach(() => {
        track = {
            download_url: '',
            attributes: {
                origin: 'external',
            },
        };
    });

    it('renders just the toggle button if showDock is false', () => {
        const component = render(
            <Dock
                showDock={false}
                track={track}
                playbackRate={1}
                triggerClassName="san-marino-blue"
                className="speed-yellow"
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders the full dock if showDock is true', () => {
        const component = render(
            <Dock
                showDock
                track={track}
                playbackRate={1}
                triggerClassName="san-marino-blue"
                className="speed-yellow"
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders the download button if the current track is downloadable', () => {
        track.download_url = 'http://download.me/track-cool.mp3';
        const component = render(
            <Dock
                showDock
                track={track}
                playbackRate={1}
                triggerClassName="san-marino-blue"
                className="speed-yellow"
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
