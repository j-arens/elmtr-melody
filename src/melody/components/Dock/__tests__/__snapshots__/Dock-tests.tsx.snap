// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dock renders just the toggle button if showDock is false 1`] = `
<div class="dock speed-yellow">
  <button class="btn dock__toggle san-marino-blue" role="button">
    <svg class="dock__toggleIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
      <path d="M19.43 12.98c.04-.32.07-.64.07-.98s-.03-.66-.07-.98l2.11-1.65c.19-.15.24-.42.12-.64l-2-3.46c-.12-.22-.39-.3-.61-.22l-2.49 1c-.52-.4-1.08-.73-1.69-.98l-.38-2.65C14.46 2.18 14.25 2 14 2h-4c-.25 0-.46.18-.49.42l-.38 2.65c-.61.25-1.17.59-1.69.98l-2.49-1c-.23-.09-.49 0-.61.22l-2 3.46c-.13.22-.07.49.12.64l2.11 1.65c-.04.32-.07.65-.07.98s.03.66.07.98l-2.11 1.65c-.19.15-.24.42-.12.64l2 3.46c.12.22.39.3.61.22l2.49-1c.52.4 1.08.73 1.69.98l.38 2.65c.03.24.24.42.49.42h4c.25 0 .46-.18.49-.42l.38-2.65c.61-.25 1.17-.59 1.69-.98l2.49 1c.23.09.49 0 .61-.22l2-3.46c.12-.22.07-.49-.12-.64l-2.11-1.65zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"></path>
    </svg>
  </button>
</div>
`;

exports[`Dock renders the download button if the current track is downloadable 1`] = `
<div class="dock speed-yellow">
  <button class="btn dock__toggle san-marino-blue" role="button">
    <svg class="dock__toggleIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
      <path d="M19.43 12.98c.04-.32.07-.64.07-.98s-.03-.66-.07-.98l2.11-1.65c.19-.15.24-.42.12-.64l-2-3.46c-.12-.22-.39-.3-.61-.22l-2.49 1c-.52-.4-1.08-.73-1.69-.98l-.38-2.65C14.46 2.18 14.25 2 14 2h-4c-.25 0-.46.18-.49.42l-.38 2.65c-.61.25-1.17.59-1.69.98l-2.49-1c-.23-.09-.49 0-.61.22l-2 3.46c-.13.22-.07.49.12.64l2.11 1.65c-.04.32-.07.65-.07.98s.03.66.07.98l-2.11 1.65c-.19.15-.24.42-.12.64l2 3.46c.12.22.39.3.61.22l2.49-1c.52.4 1.08.73 1.69.98l.38 2.65c.03.24.24.42.49.42h4c.25 0 .46-.18.49-.42l.38-2.65c.61-.25 1.17-.59 1.69-.98l2.49 1c.23.09.49 0 .61-.22l2-3.46c.12-.22.07-.49-.12-.64l-2.11-1.65zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"></path>
    </svg>
  </button>
  <span class="dock__arrow"></span>
  <div class="dock__controlsGroup">
    <button class="btn dock__control" role="button">
      <svg class="dock__controlIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
      </svg>
      <span class="dock__controlName">Download</span>
    </button>
    <button class="btn dock__control" role="button">
      <svg class="dock__controlIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"></path>
      </svg>
      <span class="dock__controlName">Speed up</span>
    </button>
    <button class="btn dock__control" role="button">
      <svg class="dock__controlIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path d="M19 13H5v-2h14v2z"></path>
      </svg>
      <span class="dock__controlName">Slow down</span>
    </button>
  </div>
</div>
`;

exports[`Dock renders the full dock if showDock is true 1`] = `
<div class="dock speed-yellow">
  <button class="btn dock__toggle san-marino-blue" role="button">
    <svg class="dock__toggleIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
      <path d="M19.43 12.98c.04-.32.07-.64.07-.98s-.03-.66-.07-.98l2.11-1.65c.19-.15.24-.42.12-.64l-2-3.46c-.12-.22-.39-.3-.61-.22l-2.49 1c-.52-.4-1.08-.73-1.69-.98l-.38-2.65C14.46 2.18 14.25 2 14 2h-4c-.25 0-.46.18-.49.42l-.38 2.65c-.61.25-1.17.59-1.69.98l-2.49-1c-.23-.09-.49 0-.61.22l-2 3.46c-.13.22-.07.49.12.64l2.11 1.65c-.04.32-.07.65-.07.98s.03.66.07.98l-2.11 1.65c-.19.15-.24.42-.12.64l2 3.46c.12.22.39.3.61.22l2.49-1c.52.4 1.08.73 1.69.98l.38 2.65c.03.24.24.42.49.42h4c.25 0 .46-.18.49-.42l.38-2.65c.61-.25 1.17-.59 1.69-.98l2.49 1c.23.09.49 0 .61-.22l2-3.46c.12-.22.07-.49-.12-.64l-2.11-1.65zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"></path>
    </svg>
  </button>
  <span class="dock__arrow"></span>
  <div class="dock__controlsGroup">
    <button class="btn dock__control" role="button">
      <svg class="dock__controlIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"></path>
      </svg>
      <span class="dock__controlName">Speed up</span>
    </button>
    <button class="btn dock__control" role="button">
      <svg class="dock__controlIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path d="M19 13H5v-2h14v2z"></path>
      </svg>
      <span class="dock__controlName">Slow down</span>
    </button>
  </div>
</div>
`;
